cluster-provision/images/vm-image-builder/create-containerdisk.sh

```diff
@@ -0,0 +1,60 @@
+#!/usr/bin/env bash
+set -exuo pipefail
+
+SCRIPT_PATH=$(dirname "$(realpath "$0")")
+
+export CUSTOMIZE_IMAGE_SCRIPT=${CUSTOMIZE_IMAGE_SCRIPT:-"${SCRIPT_PATH}/customize-image.sh"}
+
+function customize_image() {
+  local source_image=$1
+  local os_variant=$2
+  local customized_image=$3
+  local cloud_config=$4
+
+  # Backup the VM image and pass copy of the original image
+  # in case customizing script fail.
+  vm_image_copy="$(dirname $customized_image)/copy-${source_image}"
+  cp "$source_image" "$vm_image_copy"
+
+  # TODO: convert this script and its dependencies to container
+  ${CUSTOMIZE_IMAGE_SCRIPT} "$vm_image_copy" "$os_variant" "$customized_image" "$cloud_config"
+
+  # Backup no longer needed.
+  rm -f "$vm_image_copy"
+}
+
+function cleanup() {
+    if [ $? -ne 0 ];then
+      rm -rf "${build_directory}"
+    fi
+
+    rm -f "copy-${VM_IMAGE}"
+}
+
+export IMAGE_NAME=${IMAGE_NAME:-example-fedora}
+export TAG=${TAG:-32}
+export OS_VARIANT=${OS_VARIANT:-fedora31}
+export CLOUD_CONFIG_PATH=${CLOUD_CONFIG_PATH:-"${SCRIPT_PATH}/example/cloud-config"}
+export VM_IMAGE_URL=${VM_IMAGE_URL:-$(cat "${SCRIPT_PATH}/example/image-url")}
+
+readonly VM_IMAGE="source-image.qcow2"
+readonly build_directory="${IMAGE_NAME}_build"
+readonly new_vm_image_name="provisioned-image.qcow2"
+
+trap 'cleanup' EXIT SIGINT
+
+pushd "${SCRIPT_PATH}"
```

Comment:
please use '(' instead of the pushd + popd

Reply:
Using `( ` `)` cuses all command inside to run in a subshell and it is not necessay in this script
Also I prefer using `pushd` and `popd` because it is more verbose commands

Reply:
i dont see any reason to not use a sub shell,
its more nature here to use '(' without the push pop, and the ident will be more natural, now its not natural imo.
lets see what other reviewers think please.

Reply:
By verbose I meant to the commands name, so instead of `(` we use `pushd` it make it more readable in my opinion.
If you dont mind lets wait for other reviewers

Reply:
First time i had this dilemma i also thought like that, but then was asked to use `(`
and got used to it, we are adaptive, and now its self explanatory, i believe we should stick to a common practice,
it will save us precious time.

Reply:
BTW I tried it and since using `( )` creating subshell, it is necessary to catch `ERR` interupts with `trap` and adding `-E` flag in order to invoke `cleanup` on errors
Maybe this explains why you saw scripts with `-e` and `trap` that wont catch errors

Reply:
thx, ok lets then leave it as you did, just i am not sure the ident there is right to use.